src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let current = { ...state };
+  const history = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        current = {};
+        break;
+
+      case 'addProperties':
+        current = { ...current, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        current = { ...current };
+
+        action.keysToRemove.forEach((key) => {
+          delete current[key];
+        });
```

Comment:
Ensure that `action.keysToRemove` is defined and is an array before calling `forEach` on it. This will prevent potential runtime errors if the action object does not have the expected structure.